### ISSUE TITLE
atomicCompSwap(GLSL) conversion issue

### ISSUE THREAD

**Issue by yozhijk** [KhronosGroup/MoltenVK#152]
My bottom-up binary tree traversal kernel (GLSL) is using atomicCompSwap function to disable one of the threads reaching the node after handling  its kids. The kernel works fine on native AMD and NV Vulkan implementations, but produces incorrect results on MoltenVK.

The following is the problematic part of the kernel:

```
while (addr != INVALID_ADDR)
{
	// Check node's update flag.
	if (atomicCompSwap(g_tree[addr].update_flag, 0, 1) == 1)
	{
		// If the flag == 1, then the second child has already finished,
		// and this thread can do the work depending on both kids to be ready.

	}
	else
	{
                // This thread is the first reaching the node, so it bails out and let's
                // the second one handle the stuff.
		break;
	}

	addr = g_tree[addr].addr_parent;
}

```
Out of curiosity I have inspected the output of MoltenVKShaderConverter for my kernel and this is what has been produced for the code above:

```
while (addr != 4294967295u)
{
        uint _226 = 0;
        uint _229 = atomic_compare_exchange_weak_explicit(
                           (volatile device atomic_uint*)&(v_85.g_tree[addr].update_flag), 
			   &(_226), 1, 
			   memory_order_relaxed, memory_order_relaxed);

        if (_229 == 1)
        {
            // …
        }
        else
        {
            break;
        }
        addr = v_85.g_tree[addr].addr_parent;
}
```

So two observations here:

- According to MSL spec atomic_compare_exchange_weak_explicit returns bool indicating a success or a failure of the operation, not an old memory contents, so the result of _229 assignment is semantically incorrect. Old memory contents are written into _226 instead.

- C++14 standard (and I believe MSL conforms to that as well) allows weak functions to exhibit spurious failures, while GLSL atomicCompSwap provides strong success guarantees. From this point of view atomicCompSwap should be equivalent to an atomic_compare_exchange_weak called in a loop.

P.S. I was able to work around the issue using simpler atomicExchange though (which is being converted correctly).

**Comment by billhollings** [comment]
@yozhijk 

Thanks.

Can you post this exact issue to the [SPIRV-Cross issue list](https://github.com/KhronosGroup/SPIRV-Cross/issues), please?

That's where it will get fixed.

**Comment by yozhijk** [comment]
Thanks, @billhollings 

Issue created: https://github.com/KhronosGroup/SPIRV-Cross/issues/575

**Comment by billhollings** [comment]
@yozhijk Fixed in PR #159. Please retest and close.

**Comment by yozhijk** [comment]
Works fine now, thanks.
